cloud/ec2_fleet.go

```diff
@@ -0,0 +1,538 @@
+	// have access to this information now. Cost jobs will use this
+	// information later.
+	h.Zone = *instance.Placement.AvailabilityZone
+	h.StartTime = *instance.LaunchTime
+	h.VolumeTotalSize, err = getVolumeSize(ctx, m.client, h)
+	if err != nil {
+		return "", errors.Wrapf(err, "error getting volume size for host %s", h.Id)
+	}
+	if err = h.CacheHostData(); err != nil {
+		return "", errors.Wrap(err, "error updating host document in db")
+	}
+
+	// set IPv6 address, if applicable
+	for _, networkInterface := range instance.NetworkInterfaces {
+		if len(networkInterface.Ipv6Addresses) > 0 {
+			err = h.SetIPv6Address(*networkInterface.Ipv6Addresses[0].Ipv6Address)
+			if err != nil {
+				return "", errors.Wrap(err, "error setting ipv6 address")
+			}
+			break
+		}
+	}
+	return *instance.PublicDnsName, nil
+}
+
+func (m *ec2FleetManager) GetSSHOptions(h *host.Host, keyName string) ([]string, error) {
+	return getEc2SSHOptions(h, keyName)
+}
+
+func (m *ec2FleetManager) TimeTilNextPayment(h *host.Host) time.Duration {
+	return timeTilNextEC2Payment(h)
+}
+
+func (m *ec2FleetManager) getInstance(ctx context.Context, h *host.Host) (*ec2.Instance, error) {
+	instance, err := m.client.GetInstanceInfo(ctx, h.Id)
+	if err != nil {
+		grip.Error(message.WrapError(err, message.Fields{
+			"message":       "error getting instance info",
+			"host":          h.Id,
+			"host_provider": h.Distro.Provider,
+			"distro":        h.Distro.Id,
+		}))
+		return nil, errors.Wrap(err, "error getting instance info")
+	}
+
+	if err = validateEc2InstanceInfoResponse(instance); err != nil {
+		return nil, errors.Wrap(err, "invalid instance info response")
+	}
+
+	return instance, nil
+}
+
+func (m *ec2FleetManager) spawnFleetSpotHost(ctx context.Context, h *host.Host, ec2Settings *EC2ProviderSettings, blockDevices []*ec2.LaunchTemplateBlockDeviceMappingRequest) ([]*string, error) {
+	// Cleanup
+	var templateID *string
+	defer func() {
+		if templateID != nil {
+			// Delete launch template
+			_, err := m.client.DeleteLaunchTemplate(ctx, &ec2.DeleteLaunchTemplateInput{LaunchTemplateId: templateID})
+			grip.Error(message.WrapError(err, message.Fields{
+				"message":  "can't delete launch template",
+				"host":     h.Id,
+				"template": *templateID,
+			}))
+		}
+	}()
+
+	var err error
+	var templateVersion *int64
+	templateID, templateVersion, err = m.uploadLaunchTemplate(ctx, h, ec2Settings, blockDevices)
+	if err != nil {
+		return nil, errors.Wrapf(err, "unable to upload launch template for '%s'", h.Id)
+	}
+
+	instanceID, err := m.requestFleet(ctx, ec2Settings, templateID, templateVersion)
+	if err != nil {
+		return nil, errors.Wrapf(err, "can't request fleet")
+	}
+	h.Id = *instanceID
+
+	instanceInfo, err := m.client.GetInstanceInfo(ctx, h.Id)
+	if err != nil {
+		return nil, errors.Wrap(err, "can't get instance descriptions")
+	}
+
+	// return a slice of resources to be tagged
+	resources := []*string{instanceID}
```

Comment:
slices of pointers can be difficult to use, particularly ones to strings,

Reply:
changed to a slice of strings